doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,319 @@
+ * Read-Only and Read-Write interfaces
+
+Planned features:
+
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * A read-only client can talk directly, and only, to a regional read replica
+ * Discoverability of read replicas per cluster based on in-bucket configuration values
+ * Standard encoders/decoders to deliver zero-trust end-to-end encryption, perhaps using nkeys
+ * Pluggable storage backends
+
+## Data Types
+
+Here's rough guidance, for some clients in some places you might not want to use `[]string` but an iterator, that's
+fine, the languages should make appropriate choices based on this rough outline.
+
+### Entry
+
+This is the value and associated metadata made available over watchers, `Get()` etc. All backends must return an implementation providing at least this information.
+
+```go
+type Entry interface {
+	// Bucket is the bucket the data was loaded from
+	Bucket() string
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() []byte
+	// Created is the time the data was received in the bucket
+	Created() time.Time
+	// Sequence is a unique sequence for this value
+	Sequence() uint64
+	// Delta is distance from the latest value. If history is enabled this is effectively the index of the historical value, 0 for latest, 1 for most recent etc.
+	Delta() uint64
+	// Operation is the kind of operation this entry represents, enum of PUT and DEL
+	Operation() Operation
+}
+```
+
+### Status
+
+This is the status of the KV as a whole
+
+```go
+type Status interface {
+	// Bucket the name of the bucket
+	Bucket() string
+
+	// Values is how many messages are in the bucket, including historical values
+	Values() uint64
+
+	// History returns the configured history kept per key
+	History() uint64
+
+	// TTL is how long the bucket keeps values for
+	TTL() time.Duration
+
+	// BucketLocation returns a string indicating where the bucket is located, meaning dependent on backend
+	BucketLocation() string
+
+	// Keys returns a list of all keys in the bucket - not possible now
+	Keys() ([]string, error)
+
+	// BackingStore is a backend specific name for the underlying storage - eg. stream name. Used so that ops tooling can know what to Backup for example
+	BackingStore() string
+}
+```
+
+## RoKV
+
+This is a read-only KV store handle, I call this out here to demonstrate that we need to be sure to support a read-only 
+variant of the client. One that will only function against a read replica and cannot support `Put()` etc. 
+
+That capability is important, how you implement this in your language is your choice. You can throw exceptions on `Put()`
+when read-only or whatever you like.
+
+The interface here is a guide of what should function in read-only mode.
+
+```go
+// RoKV is a read-only interface to a single key-value store bucket
+type RoKV interface {
+	// Get gets a key from the store
+	Get(key string) (Entry, error)
```

Comment:
so one thing to consider is that if a key doesn't exist, we are considering that an error (which yes as far as JetStream is concerned it is) but in applications, it is NOT an error. It simply means that the value is not there. Imho the return value for go should be (*Entry, error). I know for all other JSM/etc we simply return an error, but maybe that is OK there, because really it could be a number of other things. However in this case, not finding is OK, it is just not there.

Reply:
To illustrate:
```javascript
          try {
            const ba = await this.billing.get(ac.sub);
          } catch (err) {
            //have to handle not found as an error!
            planID = "free"
            console.log(err);
          }
          // OR 
          const ba = await this.billing.get(ac.sub);
          const plan = ba ? ba.plan : "free";
```

```go
           ba, err := bucket.get(ac.sub)
           if  err != nil {
             return err;
           }
           planID := "free"
           if ba != nil {
              planID = ba.Plan
           }
```

Reply:
Entry is an interface so can’t be *Entry but it is nil on not found so I think it’s the same.